core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -591,9 +591,15 @@ private[deploy] class Master(
           true
         }
       val assignedMemory = assignedExecutors(pos) * memoryPerExecutor
-      usableWorkers(pos).memoryFree - assignedMemory >= memoryPerExecutor &&
+      val enoughMemory =
+        if (launchingNewExecutor) {
+          usableWorkers(pos).memoryFree - assignedMemory >= memoryPerExecutor
+        } else {
+          true
+        }
       usableWorkers(pos).coresFree - assignedCores(pos) >= minCoresPerExecutor &&
       coresToAssign >= minCoresPerExecutor &&
+      enoughMemory &&
       underLimit
```

Comment:
would you mind rewriting this as follows now that we have more shared code?
```
val keepScheduling = coresToAssign >= minCoresPerExecutor
val enoughCores = usableWorkers(pos).coresFree - assignedCores(pos) >= minCoresPerExecutor

// If we allow multiple executors per worker, then we can always launch new executors.
// Otherwise, if there is already an executor on this worker, just give it more cores.
val launchingNewExecutor = !oneExecutorPerWorker || assignedExecutors(pos) == 0
if (launchingNewExecutor) {
  val assignedMemory = assignedExecutors(pos) * memoryPerExecutor
  val enoughMemory = usableWorkers(pos).memoryFree - assignedMemory >= memoryPerExecutor
  val underLimit = assignedExecutors.sum + app.executors.size < app.executorLimit
  keepScheduling && enoughCores && enoughMemory && underLimit
} else {
  // We're adding cores to an existing executor, so no need to check memory and executor limits
  keepScheduling && enoughCores
}
```